aiohttp/helpers.py

```diff
@@ -741,3 +742,63 @@ def set_result(fut, result):
 def set_exception(fut, exc):
     if not fut.done():
         fut.set_exception(exc)
+
+
+class ChainedProps(Mapping):
+    __slots__ = ('_maps',)
+
+    def __init__(self, maps):
+        self._maps = tuple(maps)
+
+    def __getitem__(self, key):
+        for mapping in self._maps:
+            try:
+                return mapping[key]
+            except KeyError:
+                pass
+        raise KeyError(key)
+
+    def get(self, key, default=None):
+        return self[key] if key in self else default
+
+    def __len__(self):
+        # reuses stored hash values if possible
+        return len(set().union(*self._maps))
+
+    def __iter__(self):
+        d = {}
+        for mapping in reversed(self._maps):
+            # reuses stored hash values if possible
+            d.update(mapping)
+        return iter(d)
```

Comment:
It feels like you can have there the same `iter(set().union(*self._maps))` instead of another dict construction.

Reply:
Let's keep it for sake of shared implementation with `ChainMap`

Reply:
Iteration is not the most common call, `get`/`__getitem__` doesn't create a dict.
Moreover your proposal constructs a set, not sure what is better.